Use absolute pathes for resources, add ability to choose root conditionally

pkulev opened the pull request:
* Add readme with steps how to run the game.
* Add function for replacing local pathes with absolute ones.
* Add ability to choose root path:
** `:core-is-root t` useful for starting game after `save-lisp-and-die`
** `:core-is-root nil` is **default**. Means we want root to be ASDF system path.
    This change allows me to create bootstrap script for baking bundles, that can be easily distributed and used without installing emacs, slime, quicklisp and sbcl.
* Move resources to `data` directory. Name can be changed as you want.
    This change makes bundling easy. Just copy dir and make compressed archive!

Bootstrapping and bundling will be in separate PR.
Thanks!

pkulev committed:
Add readme

Signed-off-by: Pavel Kulyov <pkulev@croc.ru>

pkulev committed:
application: add absolute resource's path construction

Local path can be used only if image was started in project directory.
With absolute pathes we can start image anywhere.

Signed-off-by: Pavel Kulyov <pkulev@croc.ru>

alex-eg commented on application.lisp:
```diff
@@ -1,6 +1,16 @@
+
+;; TODO: provide something like config package with resource path
+;;       mappings and path-related routines. Now I can only hardcode them.
+(defmacro -res- (resource-file-name)
```
I think we can ditch dashes in `-res-`, since it's not conventional syntax, unlike `*var*` for global variables or `+const+` for global constants.

pkulev replied:
Ok, its just tempting to use some names I can't in other languages, you know.

alex-eg commented on application.lisp:
```diff
@@ -1,6 +1,16 @@
+;;       mappings and path-related routines. Now I can only hardcode them.
+(defmacro -res- (resource-file-name)
+  "Creates absolute path to resource by it's local filename."
+  `(merge-pathnames ,resource-file-name +root+))
```
If `res` intended to be a macro, why not actually merge pathnames in compile stage?
In this case, source code `(res "font3.png")` will expand to `(merge-pathnames "font3.png" +root+))`.
We claim that aark should be installed via quicklisp, so asdf should be able to find system-source-directory of aark in compile time too.
So I suggest to rewrite that macro as
```cl
(defmacro res (resource-file-name)
  "Creates absolute path to resource by it's local filename."
  (merge-pathnames resource-file-name +root+))
```

pkulev replied:
Oops, of course.

alex-eg commented on application.lisp:
```diff
@@ -1,6 +1,16 @@
+  "Frames per second.")
+
+(defparameter +root+ (asdf:system-source-directory :aark)
+  "Absolute project root path.")
```
`"Project root absolute path."`

alex-eg reviewed (changes requested):
Great patchset! Big thanks!

Some points need a little bit more attention to make everything completely perfect.

alex-eg commented on application.lisp:
```diff
@@ -4,13 +4,13 @@
   "Frames per second.")
 
 (defparameter +root+ (asdf:system-source-directory :aark)
```
I think we might need to wrap that one in `(eval-when :compile-toplevel ...)`.

pkulev committed:
Select root path conditionally

For dumping image 'core-is-root' should be t, otherwise nil.
*default-pathname-defaults* holds path where image was started, which is
the case of dumped image that can be redistributed as archive.
But in development process image can be started anywhere, so asdf helps here.

Signed-off-by: Pavel Kulyov <kulyov.pavel@gmail.com>

pkulev committed:
Move resources to make them easy bundleable

Signed-off-by: Pavel Kulyov <kulyov.pavel@gmail.com>

alex-eg commented on bootstrap.lisp:
```diff
@@ -0,0 +1,25 @@
+
+(ql:quickload :aark)
+
+#+(and (not windows) (or win32 mswindows))(pushnew :windows *features*)
```
Cool!
But I think you really should move `(pushnew  :windows *features*)` to the newline.

alex-eg commented on bootstrap.lisp:
```diff
@@ -0,0 +1,25 @@
+;;;; Creates executable for linux and windows.
+;;; usage: sbcl --load bootstrap.lisp
+;;; emacs: M-! sbcl --load bootstrap.lisp
+
```
I think you should add `(in-package :cl-user)` to the top of the file.
To be sure that all functions and variables are taken from proper package.

alex-eg commented on bootstrap.lisp:
```diff
@@ -0,0 +1,25 @@
+
+#+(and (not windows) (or win32 mswindows))(pushnew :windows *features*)
+
+(let* ((build-dir #+windows P"build/windows/" P"build/linux/")
```
Hmm... If I understand, how `#+` works, on windows this line would become
 `(let* ((build-dir P"build/windows/" P"build/linux/")`, and that doesn't look like valid let binding for me.

pkulev commented:
@alex-eg, I'm ready for review.
Please read PR description again.

alex-eg commented:
👌
